src/goto-instrument/cover_basic_blocks.cpp

```diff
@@ -0,0 +1,237 @@
+        }
+      }
+    }
+    else if(it==block_info.representative_inst)
+    {
+      // check the existing representative
+      if(!it->source_location.get_java_bytecode_index().empty())
+      {
+        if(bytecode_indices_seen.insert(
+             it->source_location.get_java_bytecode_index()).second)
+        {
+          blocks_seen.insert(block_nr);
+        }
+        else
+        {
+          // clash, reset to search for a new one
+          block_info.representative_inst=goto_program.instructions.end();
+          block_info.source_location=source_locationt::nil();
+          msg.debug() << it->function
+                      << " block " << (block_nr+1)
+                      << ", location " << it->location_number
+                      << ": bytecode-index "
+                      << it->source_location.get_java_bytecode_index()
+                      << " already instrumented."
+                      << " Searching for alternative instruction"
+                      << " to instrument." << messaget::eom;
+        }
+      }
+    }
+  }
+}
+
+void cover_basic_blockst::report_block_anomalies(
+  const goto_programt &goto_program,
+  message_handlert &message_handler)
+{
+  messaget msg(message_handler);
+  std::set<unsigned> blocks_seen;
+  forall_goto_program_instructions(it, goto_program)
+  {
+    unsigned block_nr=block_of(it);
+    const block_infot &block_info=block_infos.at(block_nr);
+
+    if(blocks_seen.insert(block_nr).second &&
+       block_info.representative_inst==goto_program.instructions.end())
+    {
+      msg.warning() << "Ignoring block " << (block_nr+1) << " location "
+                    << it->location_number << " "
+                    << it->source_location
+                    << " (bytecode-index already instrumented)"
+                    << messaget::eom;
+    }
+    else if(block_info.representative_inst==it &&
+            block_info.source_location.is_nil())
+    {
+      msg.warning() << "Ignoring block " << (block_nr+1) << " location "
+                    << it->location_number << " "
+                    << it->function
+                    << " (missing source location)"
+                    << messaget::eom;
+    }
+    // The location numbers printed here are those
+    // before the coverage instrumentation.
+  }
+}
+
+void cover_basic_blockst::output(std::ostream &out) const
+{
+  for(block_mapt::const_iterator
+        b_it=block_map.begin();
+      b_it!=block_map.end();
+      b_it++)
+    out << b_it->first->source_location
+        << " -> " << b_it->second
+        << '\n';
+}
+
+void cover_basic_blockst::update_covered_lines(block_infot &block_info)
+{
+  if(block_info.source_location.is_nil())
+    return;
+
+  const auto &cover_set=block_info.lines;
+  INVARIANT(!cover_set.empty(),
+            "covered lines set must not be empty");
+  std::vector<unsigned>
+    line_list{cover_set.begin(), cover_set.end()};
```

Comment:
@reuk will no doubt know better, but is there any advantage to using init-list-style construction here (briefly suggesting a list of iterators) rather than constructor notation (`line_list(cover_set.begin(), cover_set.end())`)?

Reply:
Basically no difference, but I prefer the curly brace syntax as it allows constructor calls to be distinguished from free function calls.

Reply:
Ahh, I always expect something behaving like aggregate init (e.g. `std::vector<int> {1,2,3}`) when I see braced init. Matter of taste I suppose, but we should arrive at some convention for the codebase and exercise it consistently.